README.md
<div id="top"></div>
<div align="center">

# RsaCtfTool

</div>

<div align=center>

[![Test](https://github.com/RsaCtfTool/RsaCtfTool/actions/workflows/test.yml/badge.svg)](https://github.com/RsaCtfTool/RsaCtfTool/actions/workflows/test.yml)
![lint_python](https://github.com/RsaCtfTool/RsaCtfTool/workflows/lint_python/badge.svg)
![CodeQL](https://github.com/RsaCtfTool/RsaCtfTool/workflows/CodeQL/badge.svg)
[![GitHub issues](https://img.shields.io/github/issues/RsaCtfTool/RsaCtfTool.svg)](https://github.com/RsaCtfTool/RsaCtfTool/issues)

</div>
<div align=center>

[![GitHub forks](https://img.shields.io/github/forks/RsaCtfTool/RsaCtfTool.svg)](https://github.com/RsaCtfTool/RsaCtfTool/network)
[![GitHub stars](https://img.shields.io/github/stars/RsaCtfTool/RsaCtfTool.svg)](https://github.com/RsaCtfTool/RsaCtfTool/stargazers)
[![GitHub license](https://img.shields.io/github/license/RsaCtfTool/RsaCtfTool.svg)](https://github.com/RsaCtfTool/RsaCtfTool)
[![GitHub contributors](https://img.shields.io/github/contributors/RsaCtfTool/RsaCtfTool.svg)](https://github.com/RsaCtfTool/RsaCtfTool/contributors)

</div>

The RSA Multi-Attack Tool is a sophisticated utility designed to decrypt data from weak public keys and attempt to recover the corresponding private key. This tool offers a comprehensive range of attack options, enabling users to apply various strategies to crack the encryption.
RSA security, at its core, relies on the complexity of the integer factorization problem. This project serves as a valuable resource by combining multiple integer factorization algorithms, effectively enhancing the overall decryption capabilities.
Please note that this tool is primarily intended for educational purposes. It is essential to manage your expectations, as not every key can be broken within a reasonable timeframe. The complexity of the encryption algorithm may present significant challenges.
It is essential to highlight that the tool, exclusively supports the RSA textbook semiprime composite modulus rather than composite multiprimes. This constraint is embedded upstream in the pycrypto library (see TODO). While this limitation exists, the tool still offers a powerful set of features for attacking RSA keys with semiprime composite modulus.

For an advanced integer factorization tool please use [msieve](https://github.com/radii/msieve), [yafu](https://github.com/bbuhrow/yafu.git), or [cado-nfs](https://gitlab.inria.fr/cado-nfs/cado-nfs).

This tool is meant for educational purposes. For those participating in CTFs, please do the following first:
* Learn the basics of RSA math and understand number theory, modular arithmetic, integer factorization and the fundamental theorem of arithmetic.
* Read the code in this repository to understand what it does and how it works and suggest improvements by sending pull requests.
* Avoid copy-pasting and running the tool without understanding the underlying math, as knowing the math is more valuable than knowing how to run the tool.

We hope this tool enhances your understanding of RSA encryption and serves as a valuable resource for exploring the intricacies of integer factorization. Use it responsibly and within the bounds of applicable laws and regulations.
                                             
__Attacks provided:__

- Attacks that don't depend on the factorization of integers (may depend on knowing n,e,ciphertext,etc...):
  - [Wiener's attack](https://en.wikipedia.org/wiki/Wiener%27s_attack)
  - Hastad's attack (Small public exponent)
  - [Boneh Durfee Method when the private exponent d is too small compared to the modulus (i.e., d < n^0.292)](https://staff.emu.edu.tr/alexanderchefranov/Documents/CMSE491/Fall2019/BonehIEEETIT2000%20Cryptanalysis%20of%20RSA.pdf)
  - Same n, huge e
  - [Small CRT exponent](https://en.wikipedia.org/wiki/Chinese_remainder_theorem)
  - Common factor between ciphertext and modulus
  - Partial q
  - Partial d
  - [Simple lattice reduction](https://en.wikipedia.org/wiki/Lattice_reduction)

- Strict Integer factorization methods (only depends on knowing n):
  - Weak public key factorization
  - Small q (q < 100,000)
  - [Fermat's factorization for close p and q](https://en.wikipedia.org/wiki/Fermat%27s_factorization_method)
  - Gimmicky Primes method
  - Past CTF Primes method
  - Non-RSA key in the form b^x, where b is the prime
  - Common factor attacks across multiple keys
  - Small fractions method when p/q is close to a small fraction
  - [Elliptic Curve Method](https://en.wikipedia.org/wiki/Lenstra_elliptic-curve_factorization)
  - [Pollards p-1 for relatively smooth numbers](https://en.wikipedia.org/wiki/Pollard%27s_p_%E2%88%92_1_algorithm)
  - Mersenne primes factorization
  - [Factordb](http://factordb.com/)
  - [Londahl](https://web.archive.org/web/20220525193825/https://grocid.net/2017/09/16/finding-close-prime-factorizations/)
  - Noveltyprimes
  - [Qicheng](https://www.cs.ou.edu/~qcheng/paper/speint.pdf)
  - binary polynomial factoring
  - [Euler method](https://en.wikipedia.org/wiki/Euler_method)
  - [Pollard Rho](https://en.wikipedia.org/wiki/Pollard%27s_rho_algorithm)
  - [Wolfram alpha](https://www.wolframalpha.com/)
  - [Z3 theorem prover](https://en.wikipedia.org/wiki/Z3_Theorem_Prover)
  - [Primorial pm1 gcd](https://en.wikipedia.org/wiki/Primorial)
  - [Mersenne Numbers pm1 gcd](https://en.wikipedia.org/wiki/Mersenne_prime)
  - [Factorial pm1 gcd](https://en.wikipedia.org/wiki/Factorial)
  - [Compositorial pm1 gcd](https://oeis.org/wiki/Compositorial)
  - [Fermat Numbers gcd](https://en.wikipedia.org/wiki/Fermat_number)
  - [Fibonacci Numbers gcd](https://en.wikipedia.org/wiki/Fibonacci_sequence)
  - System primes gcd
  - [Shanks's square forms factorization (SQUFOF)](https://en.wikipedia.org/wiki/Shanks%27s_square_forms_factorization)
  - [Return of Coppersmith's Attack (ROCA) with NECA variant](https://en.wikipedia.org/wiki/ROCA_vulnerability)
  - [Dixon](https://en.wikipedia.org/wiki/Dixon%27s_factorization_method)
  - brent (Pollard rho variant)
  - [Pisano Period](https://en.wikipedia.org/wiki/Pisano_period)
  - XYXZ form integer factorization where P prime > X^Y and Q prime > X^Z
  - High and Low Bits Equal
  - [Williams p+1](https://en.wikipedia.org/wiki/Williams%27s_p_%2B_1_algorithm)
  - [Hart algorithm (similar to Fermat)](http://wrap.warwick.ac.uk/54707/1/WRAP_Hart_S1446788712000146a.pdf)
  - [Lehmer machine (similar to Fermat)](https://en.wikipedia.org/wiki/Lehmer_sieve)
  - 2PN special form where P is prime > 2 and sqrt(2PN) is close to (Pp + 2q)/2 
  - [Kraitchik algorithm (an improvement over Fermat)](https://en.wikipedia.org/wiki/Fermat%27s_factorization_method)
  - Lehman algorithm improvement over Fermat
  - Carmichael algorithm
  - [Quadratic sieve](https://en.wikipedia.org/wiki/Quadratic_sieve)
  - [Classical part of Shor algorithm](https://en.wikipedia.org/wiki/Shor%27s_algorithm)
  - [Lucas Numbers gcd](https://en.wikipedia.org/wiki/Lucas_number)
  - [Rapid7 gcd prime dataset](https://opendata.rapid7.com/sonar.ssl/)

## Usage

```bash
usage: RsaCtfTool.py [-h] [--publickey PUBLICKEY] [--output OUTPUT] [--timeout TIMEOUT] [--createpub] [--dumpkey] [--ext] [--decryptfile DECRYPTFILE] [--decrypt DECRYPT]
                     [--verbosity {CRITICAL,ERROR,WARNING,DEBUG,INFO}] [--private] [--tests] [--ecmdigits ECMDIGITS] [-n N] [-p P] [-q Q] [-e E] [--key KEY]
                     [--password PASSWORD] [--show-factors SHOW_FACTORS]
                     [--attack {SQUFOF,XYXZ,binary_polinomial_factoring,brent,comfact_cn,cube_root,ecm,ecm2,factordb,fermat_numbers_gcd,fibonacci_gcd,highandlowbitsequal,mersenne_pm1_gcd,mersenne_primes,neca,nonRSA,noveltyprimes,pastctfprimes,pisano_period,pollard_p_1,primorial_pm1_gcd,qicheng,roca,siqs,small_crt_exp,smallfraction,smallq,system_primes_gcd,wolframalpha,wiener,boneh_durfee,euler,pollard_rho,williams_pp1,partial_q,partial_d,londahl,z3_solver,dixon,lehmer,fermat,hart,common_factors,common_modulus,same_n_huge_e,hastads,lattice,lehman,carmichael,qs,classical_shor,all} [{SQUFOF,XYXZ,binary_polinomial_factoring,brent,comfact_cn,cube_root,ecm,ecm2,factordb,fermat_numbers_gcd,fibonacci_gcd,highandlowbitsequal,mersenne_pm1_gcd,mersenne_primes,neca,nonRSA,noveltyprimes,pastctfprimes,pisano_period,pollard_p_1,primorial_pm1_gcd,qicheng,roca,siqs,small_crt_exp,smallfraction,smallq,system_primes_gcd,wolframalpha,wiener,boneh_durfee,euler,pollard_rho,williams_pp1,partial_q,partial_d,londahl,z3_solver,dixon,lehmer,fermat,hart,common_factors,common_modulus,same_n_huge_e,hastads,lattice,lehman,carmichael,qs,classical_shor,factorial_pm1_gcd,lucas_gcd,all} ...]]
                     [--sendtofdb] [--isconspicuous] [--isroca] [--convert_idrsa_pub] [--check_publickey] [--partial]
```


Mode 1 : Attack RSA (specify --publickey or n and e)

- publickey : public rsa key to crack. You can import multiple public keys with wildcards.
- decrypt : cipher message to decrypt
- private : display private rsa key if recovered

Mode 2 : Create a Public Key File Given n and e (specify --createpub)

- n : modulus
- e : public exponent

Mode 3 : Dump the public and/or private numbers (optionally including CRT parameters in extended mode) from a PEM/DER format public or private key (specify --dumpkey)

- key : the public or private key in PEM or DER format

### Decrypt file

`./RsaCtfTool.py --publickey ./key.pub --decryptfile ./ciphered\_file`

### Print private key

`./RsaCtfTool.py --publickey ./key.pub --private`

### Attempt to break multiple public keys with common factor attacks or individually- use quotes around wildcards to stop bash expansion

`./RsaCtfTool.py --publickey "*.pub" --private`


### Optionally send the results back to factordb

`./RsaCtfTool.py --publickey "*.pub" --private --sendtofdb`

### Generate a public key

`./RsaCtfTool.py --createpub -n 7828374823761928712873129873981723...12837182 -e 65537`

### Dump the parameters from a key

`./RsaCtfTool.py --dumpkey --key ./key.pub`

### Check a given private key for conspicuousness 

`./RsaCtfTool.py --key examples/conspicuous.priv --isconspicuous`

### Factor with ECM when you know the approximate length in digits of a prime

`./RsaCtfTool.py --publickey key.pub --ecmdigits 25 --verbose --private`

For more examples, look at the test.sh file

### Attack private keys with partial bits of Q known ###

`./RsaCtfTool.py --attack partial_q --key examples/masked.pem`

### Attack private keys with partial bits of D known ###

`./RsaCtfTool.py --attack partial_d --key examples/partial_d.pem`

### Convert idrsa.pub to pem format

`./RsaCtfTool.py  --convert_idrsa_pub --publickey $HOME/.ssh/id_rsa.pub`


### Check if a given key or keys are roca ###

`./RsaCtfTool.py --isroca --publickey "examples/*.pub"`

### Docker run ###

```bash
docker pull rsactftool/rsactftool
docker run -it --rm -v $PWD:/data rsactftool/rsactftool <arguments>
```


### Virtual environment run ###

Setup the venv
```bash
virtualenv venv
source venv/bin/activate
pip3 install -r requirements.txt
```

Run
```bash
source venv/bin/activate
./RsaCtfTool.py <arguments>
```

## Requirements

- python3.9
- GMPY2
- PyCrypto
- Requests
- Libnum
- SageMath : optional but advisable
- Sage binaries

### Ubuntu 18.04 and Kali specific Instructions

```bash
git clone https://github.com/RsaCtfTool/RsaCtfTool.git
sudo apt-get install libgmp3-dev libmpc-dev
cd RsaCtfTool
pip3 install -r "requirements.txt"
./RsaCtfTool.py
```


### Fedora (33 and above) specific Instructions
```bash
git clone https://github.com/RsaCtfTool/RsaCtfTool.git
sudo dnf install gcc python3-devel python3-pip python3-wheel gmp-devel mpfr-devel libmpc-devel
cd RsaCtfTool
pip3 install -r "requirements.txt"
./RsaCtfTool.py
```

If you also want the optional SageMath , you need to do
```bash
sudo dnf install sagemath
pip3 install -r "optional-requirements.txt"
```

### MacOS-specific Instructions

If `pip3 install -r "requirements.txt"` fails to install requirements accessible within the environment, the following command may work.

``easy_install `cat requirements.txt` ``

If you installed gmpy2 with homebrew(`brew install gmp`), you might have to point clang towards the header files with this command:
``CFLAGS=-I/opt/homebrew/include LDFLAGS=-L/opt/homebrew/lib pip3 install -r requirements.txt``

### Optional to factor roca keys upto 512 bits, Install neca:
You can follow the instructions at : `https://www.mersenneforum.org/showthread.php?t=23087`

## TODO (aka. Help wanted !)

- Implement a test method for each attack.
- Assign the correct algorithm complexity in **Big O** notation for each attack.
- Support multiprime RSA, the project currently supports textbook RSA.

## Contributing 

- Please read the CONTRIBUTING.md guideline for the bare minimum acceptable PRs.

## All Thanks To Our Contributors 

<a href="https://github.com/RsaCtfTool/RsaCtfTool/graphs/contributors">
  <img src="https://contrib.rocks/image?repo=RsaCtfTool/RsaCtfTool" />
</a>

<p align="right"><a href="#top">🔼 Back to top</a></p>


CHANGELOG.md
# Changelog

## Commit from 14/04/2020

### Massive code reorganization and additions

- Attacks against single key are stored in  `/attacks/single_key` folder;
- Attacks against multiple keys are stored in  `/attacks/multi_key` folder;
- Attacks are dynamically loaded;
- RsaCtfTool core is moved into `/lib` folder;
- Sage scripts are moved in `/sage` folder;
- RsaCtfTool.py is just an cli parser / some output implemented. No more logic inside;
- Multi ciphers support;
- black formatting;
- New attacks : londahl and qicheng;
- `--dumpkey` and `--ext` now dump information about each keys (public and private).

### New tests in test.sh

- Hastads;
- Decrypt multiple files;
- Testing multiple keys against on cipher;
- Extra information output.

### Fixes

- Multiple attacks are fixed;
- Better exceptions handling;
- Code cleaning;
- Fix a lot of tests.

### How to add new attacks

New RsaCtfTool architecture improve code readability and extensibility.

#### Adding single key attack

If you want to add an attack against a single key, just add a new file into `/attacks/single_key/attack_name.py` with an attack method.
Example :

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def attack(attack_rsa_obj, publickey, cipher=[]):
    # Code logic here
    return (None, None)
```

- attack_rsa_obj : a reference to `lib.rsa_attack.RSAAttack` instance. it gave access to cli arguments and various information;
- publickey : a `lib.keys_wrapper.PublicKey` instance;
- cipher : a list containing each data to decrypt.

The attack method return a tuple :
(`private_key`, `decrypted_data`)

- private_key : an instance of `lib.keys_wrapper.PrivateKey` if recovered, else None. If multiple keys are recovered, return a `list` of `lib.keys_wrapper.PrivateKey`;
- decrypted_data : if the attack focus on decrypting and not private key retrieving. If the private key is recovered, just return `None` and it will be decrypted later in script. If no data is recovered, return `None`. If multiple ciphers are recovered, return a `list` of `bytes`.

#### Adding multiple keys attack

If you want to add an attack against multiple keys, just add a new file into `/attacks/multi_key/attack_name.py` with an attack method.
Example :

```python
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

def attack(attack_rsa_obj, publickeys, cipher=[]):
    # Code logic here
    return (None, None)
```

- attack_rsa_obj : a reference to `lib.rsa_attack.RSAAttack` instance. it gave access to cli arguments and various information;
- publickeys : a list of `lib.keys_wrapper.PublicKey` instance;
- cipher : a list containing each data to decrypt.

The attack method return a tuple :
(`private_key`, `decrypted_data`)

- private_key : an instance of `lib.keys_wrapper.PrivateKey` if recovered, else None. If multiple keys are recovered, return a `list` of `lib.keys_wrapper.PrivateKey`;
- decrypted_data : if the attack focus on decrypting and not private key retrieving. If the private key is recovered, just return `None` and it will be decrypted later in script. If no data is recovered, return `None`. If multiple ciphers are recovered, return a `list` of `bytes`.

## Ending note

Still beerware licensed :)
Test, make pull requests (and issues) and add new attacks !
Keep safe & stay at home folks !


primefac.py
#! /usr/bin/env python

from __future__ import print_function, division
from threading import Timer
import _primefac
from _primefac._arith import gcd
from _primefac._prime import primes
from _primefac._util import listprod

# Note that the multiprocing incurs relatively significant overhead.
# Only call this if n is proving difficult to factor.

def kill_procs(procs):
    for p in procs:
        p.terminate()

def multifactor(n, methods=(_primefac.pollardRho_brent, _primefac.pollard_pm1, _primefac.williams_pp1,
                _primefac.ecm, _primefac.mpqs, _primefac.fermat, _primefac.factordb), verbose=False, timeout=59):
    from multiprocessing import Process, Queue as mpQueue
    def factory(method, n, output):
        try:
            g = method(n)
        except OverflowError:
            return None
        if g is not None:
          output.put((g, str(method).split()[1]))
    factors = mpQueue()

    procs = [Process(target=factory, args=(m, n, factors)) for m in methods]
    timer = Timer(timeout, kill_procs, [procs])
    try:
        timer.start()
        for p in procs:
            p.start()

        (f, g) = factors.get()
        for p in procs:
            try:
                p.terminate()
            except:
                pass
    finally:
        timer.cancel()

    if verbose:
        names = {"pollardRho_brent": "prb",
                 "pollard_pm1": "p-1",
                 "williams_pp1": "p+1"}
        if g in names:
            name = names[g]
        else:
            name = g
        print("\033[1;31m" + name + "\033[;m", end=' ')
        stdout.flush()
    return f

'''
Obtains a complete factorization of n, yielding the prime factors as they are
obtained. If the user explicitly specifies a splitting method, use that method.
Otherwise,
1.  Pull out small factors with trial division.
2.  Do a few rounds of _primefac.pollard's Rho algorithm.
    TODO: a few rounds of ECM by itself?
    TODO: a certain amount of P-1?
3.  Launch multifactor on the remainder.  Multifactor has enough overhead that
    we want to be fairly sure that rho isn't likely to yield new factors soon.
    The default value of rho_rounds=42000 seems good for that but is probably
    overkill.
'''

def primefac(n, trial_limit=1000, rho_rounds=42000, verbose=False,
             methods=(_primefac.pollardRho_brent, _primefac.pollard_pm1, _primefac.williams_pp1, _primefac.ecm, _primefac.mpqs,
                      _primefac.fermat, _primefac.factordb), timeout=60):
    from _primefac import isprime, isqrt, primegen
    from random import randrange
    timeout = timeout - 1
    if n < 2:
        return
    if isprime(n):
        yield n
        return

    factors, nroot = [], isqrt(n)
    # Note that we remove factors of 2 whether the user wants to or not.
    for p in primegen():
        if n % p == 0:
            while n % p == 0:
                yield p
                n //= p
            nroot = isqrt(n)
            if isprime(n):
                yield n
                return
        if p > nroot:
            if n != 1:
                yield n
            return
        if p >= trial_limit:
            break
    if isprime(n):
        yield n
        return

    if rho_rounds == "inf":
        factors = [n]
        while len(factors) != 0:
            n = min(factors)
            factors.remove(n)
            f = _primefac.pollardRho_brent(n)
            if isprime(f):
                yield f
            else:
                factors.append(f)
            n //= f
            if isprime(n):
                yield n
            else:
                factors.append(n)
        return

    factors, difficult = [n], []
    while len(factors) != 0:
        rhocount = 0
        n = factors.pop()
        try:
            g = n
            while g == n:
                x, c, g = randrange(1, n), randrange(1, n), 1
                y = x
                while g == 1:
                    if rhocount >= rho_rounds:
                        raise Exception
                    rhocount += 1
                    x = (x**2 + c) % n
                    y = (y**2 + c) % n
                    y = (y**2 + c) % n
                    g = gcd(x-y, n)
            # We now have a nontrivial factor g of n.  If we took too long to get here, we're actually at the except statement.
            if isprime(g):
                yield g
            else:
                factors.append(g)
            n //= g
            if isprime(n):
                yield n
            else:
                factors.append(n)
        except Exception:
            difficult.append(n)  # Factoring n took too long.  We'll have multifactor chug on it.

    factors = difficult
    while len(factors) != 0:
        n = min(factors)
        factors.remove(n)
        f = multifactor(n, methods=methods, verbose=verbose, timeout=timeout)
        if isprime(f):
            yield f
        else:
            factors.append(f)
        n //= f
        if isprime(n):
            yield n
        else:
            factors.append(n)

def factorint(n, trial_limit=1000, rho_rounds=42000, methods=(_primefac.pollardRho_brent, _primefac.pollard_pm1, _primefac.williams_pp1, _primefac.ecm, _primefac.mpqs, _primefac.fermat, _primefac.factordb)):
    out = {}
    for p in primefac(n, trial_limit=trial_limit, rho_rounds=rho_rounds, methods=methods):
        out[p] = out.get(p, 0) + 1
    return out

usage = """
This is primefac-fork version 1.1.
USAGE:
    primefac [-vs|-sv] [-v|--verbose] [-s|--summary] [-t=NUM] [-r=NUM]
          [-m=[prb][,p-1][,p+1][,ecm][,mpqs][,fermat][,factordb]] rpn
    "rpn" is evaluated using integer arithmetic.  Each number that remains on
    the stack after evaluation is then factored.
    "-t" is the trial division limit.  Default == 1000.  Use "-t=inf" to use
    trial division exclusively.
    "-r" is the number of rounds of _primefac.pollard's rho algorithm to try before
    calling a factor "difficult".  Default == 42,000.  Use "-r=inf" to use
    _primefac.pollard's rho exclusively once the trial division is completed.
    If verbosity is invoked, we indicate in the output which algorithm produced
    which factors during the multifactor phase.
    If the summary flag is absent, then output is identical to the output of the
    GNU factor command, except possibly for the order of the factors and, if
    verbosity has been turned on, the annotations indicating which algorithm
    produced which factors.
    If the summary flag is present, then output is modified by adding a single
    newline between each item's output, before the first, and after the last.
    Each item's output is also modified by printing a second line of data
    summarizing the results by describing the number of decimal digits in the
    input, the number of decimal digits in each prime factor, and the factors'
    multiplicities.  For example:
    >>> user@computer:~$ primefac  -s   24 ! 1 -   7 !
    >>>
    >>> 620448401733239439359999: 991459181683 625793187653
    >>> Z24  =  P12 x P12  =  625793187653 x 991459181683
    >>>
    >>> 5040: 2 2 2 2 3 3 5 7
    >>> Z4  =  P1^4 x P1^2 x P1 x P1  =  2^4 x 3^2 x 5 x 7
    >>>
    >>> user@computer:~$
    Note that the primes in the summary lines are listed in strictly-increasing
    order, regardless of the order in which they were found.
    The single-character versions of the verbosity and summary flags may be
    combined into a single flag, "-vs" or "-sv".
    The "-m" flag controls what methods are run during the multifactor phase.
    prb and ecm can be listed repeatedly to run multiple instances of these
    methods; running multiple instances of p-1, p+1, or mpqs confers no benefit,
    so repeated listings of those methods are ignored.
    This program can also be imported into your Python scripts as a module.
DETAILS:
    Factoring: 1.  Trial divide using the primes <= the specified limit.
               2.  Run _primefac.pollard's rho algorithm on the remainder.  Declare a
                   cofactor "difficult" if it survives more than the specified
                   number of rounds of rho.
               3.  Subject each remaining cofactor to five splitting methods in
                   parallel: _primefac.pollard's rho algorithm with Brent's improvement,
                             _primefac.pollard's p-1 method,
                             _primefac.williams' p+1 method,
                             the elliptic curve method,
                             the multiple-polynomial quadratic sieve,
                             the fermat's factorization method,
                             and search known factors using factordb.
               Using the "verbose" option will cause primefac to report which of
               the various splitting methods separated which factors in stage 3.
    RPN:       The acceptable binary operators are + - * / % **.
               They all have the same meaning as they do in Python source code
               --- i.e., they are addition, subtraction, multiplication, integer
               division, remainder, and exponentiation.
               The acceptable unary operators are ! #.  They are the factorial
               and primorial, respectively.
               There are three aliases: x for *, xx for **, and p! for #.
               You may enclose the RPN expression in quotes if you so desire.
PERFORMANCE:
CREDITS:
    Not much of this code was mine from the start.
     * The MPQS code was copied mostly verbatim from
       https://codegolf.stackexchange.com/questions/8629/9088#9088
     * The functions to manipulate points in the elliptic curve method were
       copied from a reply to the Programming Praxis post at
       http://programmingpraxis.com/2010/04/23/
"""  # TODO performance, credits

def rpn(instr):
    stack = []
    for token in instr.split():
        if set(token).issubset("1234567890L"):
            stack.append(int(token.rstrip('L')))
        elif len(token) > 1 and token[0] == '-' and set(token[1:]).issubset("1234567890L"):
            stack.append(int(token))
        elif token in ('+', '-', '*', '/', '%', '**', 'x', 'xx'):   # binary operators
            b = stack.pop()
            a = stack.pop()
            if token == '+':
                res = a + b
            elif token == '-':
                res = a - b
            elif token == '*':
                res = a * b
            elif token == 'x':
                res = a * b
            elif token == '/':
                res = a / b
            elif token == '%':
                res = a % b
            elif token == '**':
                res = a ** b
            elif token == 'xx':
                res = a ** b
            stack.append(res)
        elif token in ('!', '#', 'p!'):                             # unary operators
            a = stack.pop()
            if token == '!':
                res = listprod(range(1, a+1))
            elif token == '#':
                res = listprod(primes(a+1))
            elif token == 'p!':
                res = listprod(primes(a+1))
            stack.append(res)
        else:
            raise Exception("Failed to evaluate RPN expression: not sure what to do with '{t}'.".format(t=token))
    return [_primefac.mpz(i) for i in stack]

def main(argv):
    if len(argv) == 1:
        sysexit(usage)
    rpx, tr, rr, veb, su = [], 1000, 42000, False, False
    ms = {"prb": _primefac.pollardRho_brent,
          "p-1": _primefac.pollard_pm1,
          "p+1": _primefac.williams_pp1,
          "ecm": _primefac.ecm,
          "mpqs": _primefac.mpqs,
          "fermat": _primefac.fermat,
          "factordb": _primefac.factordb}
    methods = (_primefac.pollardRho_brent, _primefac.pollard_pm1, _primefac.williams_pp1, _primefac.ecm, _primefac.mpqs, _primefac.fermat, _primefac.factordb)
    try:
        for arg in argv[1:]:
            if arg in ("-v", "--verbose"):
                veb = True
            elif arg in ("-s", "--summary"):
                su = True
            elif arg in ("-vs", "-sv"):
                veb, su = True, True
            elif arg[:3] == "-t=":
                tr = "inf" if arg[3:] == "inf" else int(arg[3:])    # Maximum number for trial division
            elif arg[:3] == "-r=":
                rr = "inf" if arg[3:] == "inf" else int(arg[3:])    # Number of rho rounds before multifactor
            elif arg[:3] == "-m=":  # methods = tuple(ms[x] for x in arg[3:].split(',') if x in ms)
                methods = []
                for x in arg[3:].split(','):
                    if x in ms:
                        if x in ("p-1", "p+1", "mpqs") and ms[x] in methods:
                            continue
                        methods.append(ms[x])
            else:
                rpx.append(arg)
        nums = rpn(' '.join(rpx))
    except Exception as e:
        sysexit("Error while parsing arguments" + str(e))
    if su:
        print()
    for n in nums:
        print("%d: " % n, end='')
        f = {}
        for p in primefac(n, trial_limit=(n if tr == "inf" else tr),
                            rho_rounds=rr, verbose=veb, methods=methods):
            f[p] = f.get(p, 0) + 1
            print(p, end=' ')
            stdout.flush()
            assert _primefac.isprime(p) and n % p == 0, (n, p)
        print()
        if su:
            print("Z%d  = " % len(str(n)), end='')
            outstr = ""
            for p in sorted(f):
                if f[p] == 1:
                    outstr += "P%d x " % len(str(p))
                else:
                    outstr += "P%d^%d x " % (len(str(p)), f[p])
            outstr = outstr[:-2] + " = "
            for p in sorted(f):
                if f[p] == 1:
                    outstr += " %d x" % p
                else:
                    outstr += " %d^%d x" % (p, f[p])
            print(outstr[:-2])
            print()

'''
main(['p', '-s',
'1489576198567193874913874619387459183543154617315437135656']) only test
'''

# TODO timeout?
if __name__ == "__main__":
    from sys import argv as arguments, stdout, exit as sysexit
    main(arguments)

'''
Fun examples:
primefac -v 1489576198567193874913874619387459183543154617315437135656
  On my system, the factor race is a bit unpredicatble on this number.
  prb, ecm, p-1, and mpqs all show up reasonably often.
primefac -v 12956921851257164598146425167654345673426523793463
  Z50 = P14 x P17 x P20 =
      24007127617807 x 28050585032291527 x 19240648901716863967.
  p-1 gets the P14 and p+1 gets the rest.
primefac -v 38 ! 1 +
    -->  Z45 = P23 x P23 = 14029308060317546154181 x 37280713718589679646221
  The MPQS (almost always) gets this one.
    Depending on the system running things,
        this can take from 10 seconds to 3 minutes.
'''
